### DISCUSSION TITLE
Cache for local domain

### DISCUSSION THREAD

**Discussion by pedrolamas** [AdguardTeam/AdGuardHome#4369] [General]
I've been noticing an odd behavior in terms of caching for local domain!

My setup consists on all Ubiquiti UniFi devices, having the local domain DHCP served by a UniFi Dream Machine (UDM) Pro.

AdGuard Home (ADH) is running in a Docker container inside a NanoPi M4V2 SBC, and I have configured the Upstream DNS servers and Reverse DNS Servers for my local domain, pointing to the UDM Pro.

Let's say I have a host called `my_machine` that is currently off. As expected, if I query my UDM Pro it will say that it can't find that host. If I query ADH for this host, it will forward the request to the UDM Pro, and then cache the result before returning to me.

Now let's say I turn on `my_machine`!

If I now query the UDM Pro, it will correctly resolve the host name, but if I try the same on the ADH, I get a cached response with no result...

Ideally, I would say that local domain stuff should be sent to the upstream server instead of using a cached value, but the main problem is that I actually don't see this cache expiring ever!

So if the host didn't exist when I asked ADH, it will not requery for it ever again... at least that has been my experience so far!

Has anyone experienced such a thing?

Is there a way I can use to disable the cache for local resolved domains (or make the TTL really low)?

**Comment by pedrolamas** [comment]
Here's an example of what I am experiencing:

Below you can see ADH resolving a local host (upstreaming the request to my UDM Pro) that I have turned off for a while now

![image](https://user-images.githubusercontent.com/85504/157070271-360b1f8f-21bf-4418-a363-1221d7379cd9.png)

Now lets say the UDM Pro lease for that host expired, here's what ADH got on the first time it upstreamed the request:

![image](https://user-images.githubusercontent.com/85504/157070571-7bddfb43-0338-4b91-a04b-0df53dbd0172.png)

From this point on, this is what ADH returns (never again upstreams the request):

![image](https://user-images.githubusercontent.com/85504/157070630-e767477f-778b-48b8-b467-3536e499abe5.png)

Even if I now turn the host back on and UDM Pro assigns a new IP, ADH will fail to resolve the host as the cached entry never seems to expire (so no new upstream request happends!)

**Comment by ainar-g** [comment]
Hello and apologies for the late response.  Is this issue still happening in the latest version?  Also, do you have optimistic cache enabled?

**Reply by pedrolamas** [reply]
Hi @ainar-g, yes the issue is still happening with the latest version (docker).

I currently have optimistic cache disabled.

For the record, this is related with #4373

**Reply by ainar-g** [reply]
Ah, right, I remember that feature request, thanks!  We'll see what we can come up with there.

**Reply by pedrolamas** [reply]
Perfect, thank you very much! 🙂

**Comment by hengwu0** [comment]
https://github.com/AdguardTeam/dnsproxy/pull/491 fixed it, according to RFC 1035: The resource records with TTL=0 must not be cached.
